detox/runners/jest/reporters/DetoxReporter.js

```diff
@@ -25,38 +29,99 @@ class DetoxReporter extends JestVerboseReporter {
+    } else {
+      super.onTestResult(test, testResult, aggregatedResult);
+    }
+
+    if (this._lastRunComplete && this._pendingTestFiles.size === 0) {
+      this._lastRunComplete.resolve(aggregatedResult);
+    }
+  }
+
   /**
-   * @param {Set<import('@jest/reporters').TestContext>} contexts
-   * @param {import('@jest/reporters').AggregatedResult} results
+   * @param {Set<import('@jest/reporters').TestContext>} testContexts
+   * @param {import('@jest/reporters').AggregatedResult} aggregatedResult
    * @returns {Promise<any> | null}
    */
-  // @ts-ignore
-  onRunComplete(contexts, results) {
-    // Both `_lastResults` and `_runCompletePromise` are used to prevent
-    // a bug in Jest, where `onRunComplete` is called multiple times when
-    // Jest runs with `--bail` and multiple workers, and `onRunComplete`
-    // is implemented as an asynchronous long-running operation.
-    this._lastResults = [contexts, results];
+  // @ts-ignore We need to use the complete signature, not the one from the base class
+  onRunComplete(testContexts, aggregatedResult) {
     if (!this._runCompletePromise) {
-      // @ts-expect-error TS2554: Expected 0 arguments, but got 2.
-      super.onRunComplete(...this._lastResults);
-      this._runCompletePromise = this._onRunComplete();
+      // Both `_lastRunComplete` and `_runCompletePromise` are used to prevent
+      // a bug in Jest, where `onRunComplete` is called multiple times when
+      // Jest runs with `--bail` and multiple workers, and `onRunComplete`
+      // is implemented as an asynchronous long-running operation.
+      this._lastRunComplete = this._pendingTestFiles.size === 0
+        ? Deferred.resolved(aggregatedResult)
+        : new Deferred();
+
+      this._runCompletePromise = this._onRunComplete(testContexts, aggregatedResult).catch(err => {
+        console.error(err);
+        throw err;
+      });
     }
 
     return this._runCompletePromise;
   }
 
-  async _onRunComplete() {
-    let results = this._lastResults[1];
-    await reportTestResults(results.testResults.map(r => ({
+  /**
+   * @param {Set<import('@jest/reporters').TestContext>} testContexts
+   * @param {import('@jest/reporters').AggregatedResult} aggregatedResult
+   * @returns {Promise<any> | null}
+   */
+  async _onRunComplete(testContexts, { numFailedTests }) {
+    const bail = this._globalConfig.bail;
+    const earlyTeardown = bail > 0 && numFailedTests >= bail;
+    if (earlyTeardown) {
+      await unsafe_conductEarlyTeardown();
+    }
+
+    const aggregatedResult = await this._lastRunComplete.promise;
+    const lostTests = aggregatedResult.numTotalTestSuites - aggregatedResult.testResults.length;
+
+    // @ts-expect-error TS2554
+    super.onRunComplete(testContexts, aggregatedResult);
+
+    if (earlyTeardown && lostTests > 0 && config.testRunner.retries > 0) {
+      console.warn(
+        'Jest aborted the test execution before all scheduled test files have been reported.\n' +
```

Comment:
TL;DR? 😅  Can this be made more concise, perhaps by replacing some of the description with the link to the github issue in Jest? 🤔